torchtext/datasets/text_classification.py

```diff
@@ -113,33 +119,61 @@ def get_vocab(self):
+                    data_select=('train', 'test')):
+    if not tokenizer:
+        tokenizer = get_tokenizer("basic_english")
+
+    if isinstance(data_select, str):
+        data_select = [data_select]
+    if not set(data_select).issubset(set(('train', 'test'))):
+        raise TypeError('data_select is not supported!')
+
     dataset_tar = download_from_url(URLS[dataset_name], root=root)
     extracted_files = extract_archive(dataset_tar)
 
+    path = {}
     for fname in extracted_files:
-        if fname.endswith('train.csv'):
-            train_csv_path = fname
-        if fname.endswith('test.csv'):
-            test_csv_path = fname
+        if fname.endswith('train.csv') and 'train' in data_select:
+            path['train'] = fname
+        if fname.endswith('test.csv') and 'test' in data_select:
+            path['test'] = fname
 
     if vocab is None:
-        logging.info('Building Vocab based on {}'.format(train_csv_path))
-        vocab = build_vocab_from_iterator(_csv_iterator(train_csv_path, ngrams))
+        if 'train' not in data_select:
+            raise TypeError("Must pass a vocab if train is not selected.")
+        logging.info('Building Vocab based on train data')
+        if dataset_name == 'IMDB':
+            data_iterator = _imdb_iterator('train', extracted_files, tokenizer, ngrams)
+        else:
+            data_iterator = _csv_iterator(path['train'], tokenizer, ngrams)
+        vocab = build_vocab_from_iterator(data_iterator)
     else:
         if not isinstance(vocab, Vocab):
             raise TypeError("Passed vocabulary is not of type Vocab")
     logging.info('Vocab has {} entries'.format(len(vocab)))
-    logging.info('Creating training data')
-    train_data, train_labels = _create_data_from_iterator(
-        vocab, _csv_iterator(train_csv_path, ngrams, yield_cls=True), include_unk)
-    logging.info('Creating testing data')
-    test_data, test_labels = _create_data_from_iterator(
-        vocab, _csv_iterator(test_csv_path, ngrams, yield_cls=True), include_unk)
-    if len(train_labels ^ test_labels) > 0:
-        raise ValueError("Training and test labels don't match")
-    return (TextClassificationDataset(vocab, train_data, train_labels),
-            TextClassificationDataset(vocab, test_data, test_labels))
+
+    data = {}
+    for item in data_select:
+        data[item] = {}
+        data[item]['data'] = []
+        data[item]['labels'] = []
+        logging.info('Creating {} data'.format(item))
+        if dataset_name == 'IMDB':
+            data_iterator = _imdb_iterator(item, extracted_files,
+                                           tokenizer, ngrams, yield_cls=True)
+        else:
+            data_iterator = _csv_iterator(path[item], tokenizer, ngrams, yield_cls=True)
+        data_iter = _create_data_from_iterator(vocab, data_iterator, removed_tokens)
+        for cls, tokens in data_iter:
+            data[item]['data'].append((torch.tensor(cls),
```

Comment:
I wonder if it's better to create and pass a single Tensor or a list of Tensors. Should be interesting to benchmark. I'd imagine torch.Tensor indexing to be more efficient over a single block of memory than over a list of small Tensors each with their own, separate memory. This should be worth a quick benchmark.

So, more concretely, instead of append a torch.tensor here you create a list of torch.tensor(cls) and torch.tensor([token...]) and then stack them together to create one, large Tensor which is then used for construction of TextClassificationDataset.

Reply:
Sure. I can do some benchmark for this issue.
Something to figure out is that each text tensor has different length. That's where `NestedTensor` could fit.

Reply:
Yes, you're right, how did I miss that.... At least with a 2D Tensor we can translate an index to a range. Could still be interesting.

Reply:
Did some benchmark:
```
import time
import torch
tensor1 = torch.rand((10, 10))
t0 = time.time()
for i in range(0, 10000):
	result = torch.mm(tensor1, tensor1)
print("Benchmark case 1 time: ", time.time() - t0)

list_tensor = [torch.rand((10))] * 10
tensor2 = torch.stack(list_tensor)
t0 = time.time()
for i in range(0, 10000):
	result = torch.mm(tensor2, tensor2)
print("Benchmark case 2 time: ", time.time() - t0)

Benchmark case 1 time:  0.1210787296295166
Benchmark case 2 time:  0.11946344375610352
```

Reply:
Your benchmark measures the amount of time a matmul takes which seems unrelated. What I suggested here is that we look at the performance of indexing into a list and retrieving a Tensor vs. indexing into a dense tensor. You're using a dense Tensor due to torch.stack in both cases, but you're not doing any indexing.